### ISSUE TITLE
Draw text stroke before text fill 

### ISSUE THREAD

**Issue by justswim** [konvajs/konva#585]
This is a feature request for text ... I'd like to be able to write text with only an outer stroke (no inner stroke).

The main issue with the current fillStroke function is that the stroke gets added to the inside of the font, which makes the font skinnier and hard to read at small sizes.

Here is an example of what Konva does (top text is no stroke, bottom is with a stroke)
![screen shot 2019-02-22 at 3 16 35 pm](https://user-images.githubusercontent.com/1372528/53276641-de1d8300-36b4-11e9-90dc-c193ff15736e.png)

If the stroke is only applied to the outside of the text, it could look like this. Note how the text is much more readable in this case:
![screen shot 2019-02-22 at 3 14 40 pm](https://user-images.githubusercontent.com/1372528/53276625-b3332f00-36b4-11e9-9291-74c301d61093.png)

Basically, I'd like to achieve the same result as a text shadow with offset, but the offset should be 360 degrees around the text.

Is this possible somehow?

Thanks for an amazing project!

**Comment by justswim** [comment]
I think the solution would be for text, to first stroke the text, then fill it. Would it be possible to make this change in Konva?

**Comment by lavrton** [comment]
At the current moment, you can't do that with `Konva.Text`. The only solution is to write a custom shape and apply stroke/fill manually.

Probably we can add a property something like `shouldFillAfterStroke`.

**Comment by justswim** [comment]
Thanks for your quick reply. Is this something that you can easily add? Or should I send a pull request?

My initial thinking is if the shouldfillafterstroke parameter is true then change the order of the functions in ctx.fillthestroke. Would that strategy work?

**Comment by lavrton** [comment]
I am not sure that is an important feature, need to think about including it into `Konva` core.

As workaround you can do this:

```javascript
const originalFillStroke = layer.getContext().fillStrokeShape;

layer.getContext().fillStrokeShape = function(shape){
  if (shape instanceof Konva.Text) {
    if (shape.getStrokeEnabled()) {
      this._stroke(shape);
    }
    if (shape.getFillEnabled()) {
      this._fill(shape);
    }
  } else {
    originalFillStroke.call(this, shape);
  }
}
```

https://jsbin.com/juxulaqebe/1/edit?html,js,output

**Comment by justswim** [comment]
@lavrton The workaround works perfectly for me so far, thank you!

**Comment by shahgahmed** [comment]
@lavrton Thank you for the guidance above.  When an opacity below 1 or a filter is applied to text, the stroke is not as expected given the workaround.  

Demo: https://galeeb.com/konva

Given the above, does it make sense to reopen the issue?

**Comment by junho85** [comment]
@lavrton 
Thanks for the workaround. I need this feature.

**Comment by junho85** [comment]
@lavrton oops. There is a problem.
If I download picture (https://konvajs.org/docs/data_and_serialization/Stage_Data_URL.html) not applied the workaround.

**Comment by lavrton** [comment]
Try to use this:
```javascript
const originalFillStroke = Konva.Context.prototype.fillStrokeShape;
Konva.Context.prototype.fillStrokeShape = function(shape) {
        if (shape instanceof Konva.Text) {
          if (shape.getStrokeEnabled()) {
            this._stroke(shape);
          }
          if (shape.getFillEnabled()) {
            this._fill(shape);
          }
        } else {
          originalFillStroke.call(this, shape);
        }
};
```

Probably we should make it a default for Text shape.

**Comment by junho85** [comment]
@lavrton 
God! Perfect!

**Comment by ByronMorley** [comment]
I would like to use this in React, is there any way I could use it. or is there a way to convert a standard kona Text to React, any help would be much appreciated

**Comment by lavrton** [comment]
The workaround above will work for React too.

**Comment by GeorgeGkas** [comment]
What is the current status of this issue? This is labeled as `enchantment` but workarounds exist. Should we close it?

**Comment by lavrton** [comment]
Good news everyone! The feature is implemented and released with `konva@7.2.0`.

For demo: https://konvajs.org/docs/styling/Fill_Stroke_Order.html

**Comment by LaureanoBlonsky** [comment]
Hi lavtron! I found a bug:
It works ok, but when you cache the layer, it renders both shadows (fill and border) together.

Fiddle: https://jsfiddle.net/laureanoblonsky/5pbnoyrf/7/

**Comment by lavrton** [comment]
@LaureanoBlonsky It is a bug and may be hard to resolve.

**Comment by LaureanoBlonsky** [comment]
Any idea how to resolve it? I can explore if you point me somewere.

Meanwhile I can just add another identical text on top without shadow, as a workaround

**Comment by lavrton** [comment]
Konva is using buffer canvas to draw some nodes off-screen before rendering on real scene canvas on the page. It is required to fix some inconsistencies and resolve issues like above with shadows.
But Konva can't use that buffer canvas while caching a node. So, we have that effect.
